### ISSUE TITLE
Integrate with go vet

### ISSUE THREAD

**Issue by ironcladlou** [ironcladlou/GoTools#113]
Add support for `go vet`.

**Comment by stevekuznetsov** [comment]
As we have learned today this seems to not be a trivial matter. Hopefully from that conversation we can learn how to best use `go vet` and answer questions like:
* how do we ensure that two people running the same version of `go vet` on the same code result in the same output?
* what are the pros and cons of running `go vet`on specific source files versus a tree or directory?
* how do we ensure that changes in `go vet` that are picked up by users don't cause bugs in this code?

**Comment by ironcladlou** [comment]
I played with `vet` more in depth today. It actually brings us to a design impasse. The tool works in the following contexts:

1. A *single* package in GOPATH
2. Named `go` files in *the same package*
3. All packages found recursing into *a named directory*

The current GoTools project settings structure doesn't quite mesh with the available `vet` options for a nontrivial case. Number 3 is closest, and can pretty much do the right thing in theory by running against `{gopath[x]}/{project_package}`; the problem there is that for nontrivial go projects it's almost guaranteed to pick up directories you don't actually want to `vet` (like vendored dependencies).

I'm not sure if a new setting (e.g.`vet_exclusions`) will be avoidable. Also have to decide whether to whitelist or blacklist.

**Comment by ironcladlou** [comment]
I was looking over vim-go for inspiration, and I think I see the way forward here. Vim-go has a `gometalinter` command which lets you run all the popular linters. I think introducing that concept in GoTools would make sense and also warrants a new settings key like:

```json
{
  "linters": {
    "vet": { "dirs": ["pkg", "test"], "filter": "<pattern>" },
    "errcheck": { }
  }
}
```

A `lint` command would run all linters (by default) or the specified linters. We have the base project path by locating `project_package` within `gopath`. If `dirs` is empty, feed the project path to the linter's recursive mode. If `dirs` is not empty, then for each dir feed the dir path to the linter's recursive mode and accumulate the results. Prune the linter output by requiring each line match `filter`. Render to a panel with an output regex smart enough to parse the output of all supported linters.

The same command could be used to lint just the current file, or the package of the current file. Need to think more on what the command API would be.

Thoughts?

**Comment by ironcladlou** [comment]
> If dirs is empty, feed the project path to the linter's recursive mode. If dirs is not empty, then for each dir feed the dir path to the linter's recursive mode and accumulate the results.

A simpler way of expressing this would be: For each directory, run the linter in recursive mode and accumulate the results. The `dirs` setting would default to just `[ <project directory> ]` which would be "all packages."

**Comment by stevekuznetsov** [comment]
Are we attempting to run this in order to do error highlighting? Why don't we just invoke it for the currently focused file?

**Comment by ironcladlou** [comment]
>Are we attempting to run this in order to do error highlighting?

I don't think so, at least not at first. I consider it more like the `oracle` command- it'll give you a plain output panel with clickable references.

>Why don't we just invoke it for the currently focused file?

That will be possible, in addition to the other modes (all packages in some/all dirs, all files in current package). I want them all.

**Comment by ironcladlou** [comment]
>>Are we attempting to run this in order to do error highlighting?

>I don't think so, at least not at first. I consider it more like the oracle command- it'll give you a plain output panel with clickable references.

To be clear, adding markers is not hard, just not my priority for v1 of the feature.

**Comment by stevekuznetsov** [comment]
I see, that make sense. A click-able panel that allows me to jump to errors would definitely be very useful. The outline you gave seems sensible to me. We should perhaps expose the different flags for `vet` (`shadows`,`composites`, etc) instead of throwing the output lines through a regex filter - unclear what the purpose of that is. Given that errors can occlude others (when `vet` gives up prematurely after finding other problems) we do not want to be invoking it indiscriminately and then ignoring some output as that could mask what we're interested in.

**Comment by ironcladlou** [comment]
Filtering might not be necessary off the bat if you can choose directories for the package scanning. Mainly I wanted to avoid running things like Godeps or generated stuff like assets blobs if possible by post-filtering the output (as not all the tools support any filtering internally for their recursive options). Being able to whitelist dirs might be enough to cover all the typical cases without needing filtering.

For adding linter-specific flags, I'd say that would be a followup.

As far as accidentally filtering errors: the tools all return nonzero on exit and also write their errors to stderr, so we know not to filter when problems occur, and also not to filter stderr at all (unless working around some idiosyncrasy of a particular tool).

Anyway, I think I've talked myself out of filtering at all for the moment.

**Comment by nuance** [comment]
This doesn't address the configuration messiness (although wiring in whatever ends up being the design shouldn't be hard), but I've built a limited version of this in https://github.com/nuance/GoTools/commit/c0736396f56f0a31600eab4f8fb903e56342a9ff
